File: 2018/julia/day09.jl
mutable struct Marble
    prev::Int
    next::Int
end

mutable struct Ring
    curr::Int
    vals::Array{Marble}
    Ring(max_size) = new(1, [Marble(1, 1)  for _ in 1:max_size + 1])
end

function insert!(self::Ring, value)
    next_value = self.vals[self.curr].next
    self.vals[value] = Marble(self.curr, next_value)
    self.vals[self.curr].next = value
    self.vals[next_value].prev = value
end

function remove!(self::Ring)
    curr, rem = self.curr, self.vals[self.curr]
    prev_value, next_value = rem.prev, rem.next
    self.vals[prev_value].next = next_value
    self.vals[next_value].prev = prev_value
    self.curr = prev_value
    return curr
end

function step!(self::Ring, steps)
    for _ in 1:steps
        self.curr = self.vals[self.curr].next
    end
end

function back!(self::Ring, steps)
    for _ in 1:steps
        self.curr = self.vals[self.curr].prev
    end
end

function compute_scores(num_players, num_marbles)
    player, scores = 0, zeros(Int, num_players)
    marbles = Ring(num_marbles)
    for marble in 1:num_marbles
        player = player % num_players + 1
        step!(marbles, 2)
        if marble % 23 == 0
            back!(marbles, 8)
            scores[player] += marble + remove!(marbles) - 1
        else
            insert!(marbles, marble + 1)
        end
    end
    return scores
end

vals = [v for line in readlines("data/input-09.txt") for v in split(line, " ")]
num_players, num_marbles = parse(Int, vals[1]), parse(Int, vals[7])

res = maximum(compute_scores(num_players, num_marbles))
println("Part 1: $(res)")
@assert(res == 385820)

res = maximum(compute_scores(num_players, num_marbles * 100))
println("Part 2: $(res)")
@assert(res == 3156297594)

File: 2018/python/day09.py
class Marble:
    def __init__(self, prev, succ):
        self.prev = prev
        self.succ = succ


class Ring:
    def __init__(self, max_size):
        self.curr = 0
        self.vals = [Marble(0, 0)] * (max_size + 1)

    def insert(self, value):
        succ_value = self.vals[self.curr].succ
        self.vals[value] = Marble(self.curr, succ_value)
        self.vals[self.curr].succ = value
        self.vals[succ_value].prev = value

    def remove(self):
        curr, rem = self.curr, self.vals[self.curr]
        prev_value, succ_value = rem.prev, rem.succ
        self.vals[prev_value].succ = succ_value
        self.vals[succ_value].prev = prev_value
        self.curr = prev_value
        return curr

    def step(self, steps):
        for _ in range(steps):
            self.curr = self.vals[self.curr].succ

    def back(self, steps):
        for _ in range(steps):
            self.curr = self.vals[self.curr].prev


def compute_scores(num_players, num_marbles):
    player, scores = 0, [0] * num_players
    marbles = Ring(num_marbles)
    for marble in range(1, num_marbles + 1):
        player = (player + 1) % num_players
        marbles.step(2)
        if marble % 23 == 0:
            marbles.back(8)
            scores[player] += marble + marbles.remove()
        else:
            marbles.insert(marble)
    return scores


file = open("data/input-09.txt")
vals = [v for line in file.readlines() for v in line.split(" ")]
num_players, num_marbles = int(vals[0]), int(vals[6])

res = max(compute_scores(num_players, num_marbles))
print(f"Part 1: {res}")
assert res == 385820

res = max(compute_scores(num_players, num_marbles * 100))
print(f"Part 2: {res}")
assert res == 3156297594
